### PR TITLE
Feature: Reusable slider & refactor sliders

### PR DESCRIPTION
**Pull request by JakubFrackowiak**
Resolves: #115

### CODE REVIEW

**Comment by piotrostr** [comment]
the green button is off on mobile

**Comment by piotrostr** [comment]
the text should not be black in the dark theme

File: components/ReusableComponents/ReusableSlider/ReusableSlider.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import { LatestQuest } from "components/LatestQuests/LatestQuest"
+import { SliderButtons } from "components/ReusableComponents/ReusableSlider/SliderButtons"
+import { SliderHeader } from "./SliderHeader"
+import { Role } from "components/Role/Role"
+import { Role as RoleType } from "storage/team"
+
+interface Item extends Hero, Team, Quest, RoleType {}
+
+interface ReusableSliderProps {
+  variant: "hero" | "team" | "quest" | "role"
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/ReusableSlider.tsx:18]: use Enums here

File: components/ReusableComponents/ReusableSlider/ReusableSlider.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}: ReusableSliderProps) {
+  const [scrolledCard, setScrolledCard] = useState(0)
+  const [mouseScrollDisabled, setMouseScrollDisabled] = useState(false)
+  const featuredCardsRefs = useRef([])
+  const featuredCardsContainerRef = useRef()
+  const isMobile = useMediaQuery("(max-width: 600px)")
+
+  const renderItem = (item: Item) => {
+    switch (variant) {
+      case "hero":
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/ReusableSlider.tsx:36]: match against enums

File: components/ReusableComponents/ReusableSlider/ReusableSlider.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import { Quest } from "storage/quest"
+import { LatestTeam } from "components/LatestTeams/LatestTeam"
+import { HeroAvatar } from "components/HeroAvatar"
+import { LatestQuest } from "components/LatestQuests/LatestQuest"
+import { SliderButtons } from "components/ReusableComponents/ReusableSlider/SliderButtons"
+import { SliderHeader } from "./SliderHeader"
+import { Role } from "components/Role/Role"
+import { Role as RoleType } from "storage/team"
+
+interface Item extends Hero, Team, Quest, RoleType {}
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/ReusableSlider.tsx:15]: if the Item extends all of those types it becomes an extremely general type, basically `any`

I would suggest to use `item = item as SomeType`

File: components/ReusableComponents/ReusableSlider/SliderButtons.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
+import Link from "next/link"
+import { Button, Stack, Typography } from "@mui/material"
+
+interface Variant {
+  variant: "hero" | "quest" | "team" | "role"
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/SliderButtons.tsx:5]: DRY

File: components/ReusableComponents/ReusableSlider/SliderHeader.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+function HeroesSliderHeader() {
+  return (
+    <Stack direction="row" justifyContent="space-between" alignItems="center">
+      <Stack spacing={2}>
+        <Typography variant="h4" fontWeight={600}>
+          Don't just take our word for it
+        </Typography>
+        <Typography variant="body1" color="text.secondary">
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/SliderHeader.tsx:14]: those could be mapped out

File: components/ReusableComponents/ReusableSlider/SliderHeader.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    </Stack>
+  )
+}
+
+export function SliderHeader({ variant }: Variant) {
+  const renderHeader = () => {
+    switch (variant) {
+      case "hero":
+        return <HeroesSliderHeader />
+      case "quest":
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/SliderHeader.tsx:73]: this is also repeated

File: components/ReusableComponents/ReusableSlider/ReusableSlider.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
+import { Box, Container, Stack, Typography, useMediaQuery } from "@mui/material"
+import { ScrollLeft, ScrollRight } from "components/ScrollButtons"
+import { useRef, useState } from "react"
+import { Hero } from "storage/hero"
+import { Team } from "storage/team"
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/ReusableSlider.tsx:5]: no absolute imports

File: components/ReusableComponents/ReusableSlider/SliderButtons.tsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+      </Link>
+    </Stack>
+  )
+}
+
+export function SliderButtons({ variant }: Variant) {
+  const renderButtons = () => {
+    switch (variant) {
+      case "hero":
+        return <HeroesSliderButtons />
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/SliderButtons.tsx:116]: no need to wrap this into a function and proc it, drop the `renderButtons`

**Review by piotrostr — CHANGES_REQUESTED** [review]

File: components/ReusableComponents/ReusableSlider/SliderButtons.tsx [hunk 9]
```diff
@@ -20,7 +20,7 @@ function HeroesSliderButtons() {
             height: "3rem",
             borderRadius: "0.5rem",
             borderColor: (theme) => theme.palette.grey[300],
-            width: "100%",
+            width: "11rem",
```
Comment (piotrostr) [inline components/ReusableComponents/ReusableSlider/SliderButtons.tsx:20]: 11rem is a bit much, use a max-width and some of sm, lg, xl etc

**Review by piotrostr — CHANGES_REQUESTED** [review]

File: components/ReusableComponents/ReusableSlider/SliderButtons.tsx (hunk 9 above)
Comment (JakubFrackowiak) [reply components/ReusableComponents/ReusableSlider/SliderButtons.tsx]: I deleted the width prop entirely from all button variants. I used the whiteSpace="nowrap" to make text always stay in one line on the button.  Using for eg. maxWidth="md" doesnt really work on mui buttons.

**Review by piotrostr — CHANGES_REQUESTED** [review]
The directory should not be named ReusableComponents, I would suggest to keep the Slider in the main components directory and there is no need to indicate it's reusability so explicitly.

File: components/LatestQuestsSlider/index.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  const latestQuestsRefs = useRef([])
+  const latestQuestsContainerRef = useRef()
+
+  const isMobile = useMediaQuery("(max-width: 600px)")
+
+  return (
+    <Stack spacing={4} alignItems="center" sx={{ overflow: "clip" }}>
+      <Container>
+        <Stack spacing={6}>
+          <Stack direction="row" justifyContent="space-between">
```
Comment (piotrostr) [inline components/LatestQuestsSlider/index.tsx]: how does this scale on mobile?

**Review by piotrostr — CHANGES_REQUESTED** [review]

File: components/LatestQuestsSlider/index.tsx (hunk 10 above)
Comment (JakubFrackowiak) [reply components/LatestQuestsSlider/index.tsx]: So accually there is no LatestQuestSlider component due to reusable one existing, I ll attach a screen of a slider on mobile

Comment (JakubFrackowiak) [reply components/LatestQuestsSlider/index.tsx]: image

image

@piotrostr

**Comment by JakubFrackowiak** [comment]
all the other sliders should be deleted, only reusable one should be left

**Comment by JakubFrackowiak** [comment]
Conflicts are resolved now @piotrostr

**Review by piotrostr — APPROVED** [review]

**Pull request merged**